server/src/storage/mod.rs

```diff
@@ -68,32 +67,37 @@ impl StorageEngine {
             Ok(data_point) => {
                 let key = data_point.generate_unique_key();
 
-                println!("Inserting data point {:?} into segment with key '{}'.", data_point, key);
+                info!("Inserting data point {:?} into segment with key '{}'.", data_point, key);
 
                 if let Some(segment) = self.data.get_mut(&key) {
-                    println!("Found existing segment with key '{}'.", key);
+                    info!("Found existing segment with key '{}'.", key);
 
                     segment.insert_data(&data_point);
 
                     if segment.is_full() {
-                        println!("Segment is full, moving it to the compression queue.");
+                        info!("Segment is full, moving it to the compression queue.");
 
                         let full_segment = self.data.remove(&key).unwrap();
                         self.enqueue_segment(key, full_segment)
                     }
                 } else {
-                    println!(
-                        "Could not find segment with key '{}'. Creating segment.",
-                        key
-                    );
+                    info!("Could not find segment with key '{}'. Creating segment.", key);
+
+                    // If there is not enough space for a new segment, spill a finished segment.
+                    if SegmentBuilder::get_memory_size() > self.remaining_bytes {
+                        self.spill_finished_segment();
+                    }
 
+                    // Create a new segment and remove the size from the reserved remaining memory.
                     let mut segment = SegmentBuilder::new();
-                    segment.insert_data(&data_point);
+                    self.remaining_bytes -= SegmentBuilder::get_memory_size();
+                    info!("Created segment. Remaining bytes: {}.", self.remaining_bytes);
```

Comment:
Maybe add `key` or use a `tracing::span` to make it clear for which time series the segment was created.

Reply:
Fixed.